### DISCUSSION TITLE
Why does DrawingView doesn't have DrawingLineStarted event or DrawingLineMoved event?

### DISCUSSION THREAD

**Discussion by Kamesuta** [CommunityToolkit/Maui#1419] [New Feature Discussions]
Hello, I would like to propose adding drawing start event and move event to DrawingView.

# Problem
- DrawingView has an event called DrawingLineCompleted event when drawing is completed, but no events are provided at the start of drawing or while drawing is in progress.
    - I tried to use some hacky(dirty) code to achieve this, but they were not possible because there was no event and inheritance was not allowed. Therefore, I was convinced that a code change was needed to accomplish that.

# Proposal
- I want to add "drawing start event" and "move cursor event" to DrawingView.

# Concerns with this proposal
- One concern is that there may be a slight performance disadvantage because an event is fired for each mouse movement, but since IDrawingView::DrawAction already does the same thing with frequency, this should not be a problem.

# What I wanted to do and what I did
- I want to make an application that synchronizes multiple DrawingViews across the network, so I need to hook into those events.
- I have checked that the event can be obtained by changing the repository/branch as follows.
  - https://github.com/Kamesuta/Maui/compare/cfc7832...047de11
- The code I want to do is something like this ↓
  ```csharp
  DrawingViewControl.DrawingLineStartedAction += (point) =>
  {
      Debug.WriteLine($"DrawingLineStarted: X={point.X}, Y={point.Y}");
  };
  DrawingViewControl.DrawingLineMovedAction += (point) =>
  {
      Debug.WriteLine($"DrawingLineMoved: X={point.X}, Y={point.Y}");
  };
  DrawingViewControl.DrawingLineCompleted += (e, args) =>
  {
      Debug.WriteLine($"DrawingLineCompleted: Points={args.LastDrawingLine.Points.Count}");
  };
  ```

**Comment by VladislavAntonyuk** [comment]
I like the idea. For now, I suggest you extend our control with your own. and add these additional actions/events
If the user is not subscribed to the event there should not be performance degradation

**Reply by Kamesuta** [reply]
> I suggest you extend our control with your own. and add these additional actions/events

Does this mean that you want me to fork and rewrite the source code to extend it? (and Pull Request it?)
Or do you have an idea that can be extended from the application side, such as inheritance?

As far as I can see from the source code, [the functions called when drawing is started/moved](https://github.com/CommunityToolkit/Maui/blob/fcf6569/src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs#L80-L122) are private access and do not seem to be extendable.
https://github.com/CommunityToolkit/Maui/blob/fcf6569/src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs#L80-L122

**Reply by pictos** [reply]
You can follow this doc to have an idea on how to it
https://learn.microsoft.com/en-us/dotnet/maui/user-interface/handlers/customize

**Comment by cat0363** [comment]
@Kamesuta , Is the following what you want to achieve?

https://github.com/CommunityToolkit/Maui/assets/125236133/6bbd9914-387f-4987-ae01-cbd41d696e4c

I modified the latest source code of main and implemented it.

To achieve this, you need to modify the following files.

[src\CommunityToolkit.Maui\Views\DrawingView\DrawingView.shared.cs]
[src\CommunityToolkit.Maui.Core\Views\DrawingView\PlatformView\MauiDrawingView.shared.cs]
[src\CommunityToolkit.Maui.Core\Handlers\DrawingView\DrawingViewHandler.shared.cs]
[src\CommunityToolkit.Maui.Core\Interfaces\IDrawingView.shared.cs]

Also, to achieve this you need to add the following files:

[src\CommunityToolkit.Maui.Core\Primitives\DrawingLineStartedEventArgs.shared.cs]
[src\CommunityToolkit.Maui.Core\Primitives\DrawingLineMovedEventArgs.shared.cs]
[src\CommunityToolkit.Maui.Core\Primitives\MauiDrawingLineStartedEventArgs.shared.cs]
[src\CommunityToolkit.Maui.Core\Primitives\MauiDrawingLineMovedEventArgs.shared.cs]

Additional information:
Android was fast, but iOS and Windows were slow. However, this has been verified through 
debugging, so it may be different in the release.

**Reply by Kamesuta** [reply]
> Is the following what you want to achieve?

Yes, it is. It is exactly the ideal behavior!

**Reply by Kamesuta** [reply]
I was presenting my implementation when I set up this discussion.

> I have checked that the event can be obtained by changing the repository/branch as follows.
> https://github.com/Kamesuta/Maui/compare/cfc7832...047de11

In my implementation, I don't have the following class and pass PointF directly to Action\<PointF\>

> Also, to achieve this you need to add the following files:
> [src\CommunityToolkit.Maui.Core\Primitives\DrawingLineStartedEventArgs.shared.cs]
> [src\CommunityToolkit.Maui.Core\Primitives\DrawingLineMovedEventArgs.shared.cs]

Would it be slow to create an instance of the DrawingLineMovedEventArgs class for each move even though no events have been registered?

**Reply by cat0363** [reply]
@Kamesuta , The PR created by @VladislavAntonyuk is almost the same as the one I created.
At least once I unregistered the event, it became lighter, so you think it's as expected.
Yesterday, I was displaying the coordinates on the Editor for debugging, but once I stopped 
doing that, it became lighter to some extent, so I think it depends on how you use it.

>Would it be slow to create an instance of the DrawingLineMovedEventArgs class for each move even though no events have been registered?

**Comment by VladislavAntonyuk** [comment]
the pr is coming today/tomorrow

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889
